x-pack/metricbeat/module/awsfargate/task_stats/cpu.go

```diff
@@ -0,0 +1,172 @@
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package task_stats
+
+import (
+	"strconv"
+
+	"github.com/docker/docker/api/types"
+
+	"github.com/elastic/beats/v7/libbeat/common"
+	"github.com/elastic/beats/v7/libbeat/logp"
+)
+
+// Stat contains container and statistics information
+type Stat struct {
+	Container *types.Container
+	Stats     types.StatsJSON
+}
+
+// cpuUsage is a struct representation of container CPU usage stats
+type cpuUsage struct {
+	*Stat
+
+	cpus        uint32
+	systemDelta uint64
+}
+
+type cpuStats struct {
+	PerCPUUsage                           common.MapStr
+	TotalUsage                            float64
+	TotalUsageNormalized                  float64
+	UsageInKernelmode                     uint64
+	UsageInKernelmodePercentage           float64
+	UsageInKernelmodePercentageNormalized float64
+	UsageInUsermode                       uint64
+	UsageInUsermodePercentage             float64
+	UsageInUsermodePercentageNormalized   float64
+	SystemUsage                           uint64
+	SystemUsagePercentage                 float64
+	SystemUsagePercentageNormalized       float64
+}
+
+func getCPUStats(taskStats types.StatsJSON) cpuStats {
+	usage := cpuUsage{Stat: &Stat{Stats: taskStats}}
+
+	return cpuStats{
+		TotalUsage:                            usage.Total(),
+		TotalUsageNormalized:                  usage.TotalNormalized(),
+		UsageInKernelmode:                     taskStats.Stats.CPUStats.CPUUsage.UsageInKernelmode,
+		UsageInKernelmodePercentage:           usage.InKernelMode(),
+		UsageInKernelmodePercentageNormalized: usage.InKernelModeNormalized(),
+		UsageInUsermode:                       taskStats.Stats.CPUStats.CPUUsage.UsageInUsermode,
+		UsageInUsermodePercentage:             usage.InUserMode(),
+		UsageInUsermodePercentageNormalized:   usage.InUserModeNormalized(),
+		SystemUsage:                           taskStats.Stats.CPUStats.SystemUsage,
+		SystemUsagePercentage:                 usage.System(),
+		SystemUsagePercentageNormalized:       usage.SystemNormalized(),
+	}
+}
+
+// CPUS returns the number of cpus. If number of cpus is equal to zero, the field will
+// be updated/initialized with the corresponding value retrieved from Docker API.
+func (u *cpuUsage) CPUs() uint32 {
+	if u.cpus == 0 {
+		if u.Stats.CPUStats.OnlineCPUs != 0 {
+			u.cpus = u.Stats.CPUStats.OnlineCPUs
+		} else {
+			//Certain versions of docker don't have `online_cpus`
+			//In addition to this, certain kernel versions will report spurious zeros from the cgroups usage_percpu
+			var realCPUCount uint32
+			for _, rCPUUsage := range u.Stats.CPUStats.CPUUsage.PercpuUsage {
+				if rCPUUsage != 0 {
+					realCPUCount++
+				}
+			}
+			u.cpus = realCPUCount
+		}
+
+	}
+	return u.cpus
+}
+
+// SystemDelta calculates system delta.
+func (u *cpuUsage) SystemDelta() uint64 {
```

Comment:
Wondering if we could somehow avoid duplicating this with https://github.com/elastic/beats/blob/949fb147336e41447c0f842d3e28d9f1073e926d/metricbeat/module/docker/cpu/helper.go#L122. Maybe by using a base interface and apply some kind of extensions?

Reply:
Actually if we make `cpuUsage` public, then we can reuse all the methods here in awsfargate. Please see the latest commit 🙂

Reply:
Aha never mind, I fixed the doc issue with https://github.com/elastic/beats/pull/22034/commits/9e61b9079c54e0dd1ca97855cf7912dbc746162e#diff-1333975e4127602a5d3ef9c0758fb8eeb86370576a49681bdadf4335f3bf93bfR161